### PR TITLE
Update packpack for building new alpine packages

### PR DESCRIPTION
**Pull request by foxzi**
Building alpine needs ssl keys for sign packages. Before this fix
generating keys were hardcoded in packpack and impossible save and use again.

Now for buildning alpine package depends on variable ABUILD_KEY. In this variable
should be stored private key.

Example show how to create alpine keys.
`docker run -it --rm alpine:3.16 sh -c 'apk add alpine-sdk; abuild-keygen -n; cat ~/.abuild/*'`

More information - https://wiki.alpinelinux.org/wiki/Abuild_and_Helpers

### CODE REVIEW

**Review by ylobankov — COMMENTED** [review]
I tried to build tarantool 2.10 packages and got the following error:

```bash
/build/src/tarantool-2.10.2/src/lib/tzcode/strptime.c:40:10: fatal error: sys/cdefs.h: No such file or directory
   40 | #include <sys/cdefs.h>
      |          ^~~~~~~~~~~~~
[  0%] No update step for 'bundled-ares-project'
compilation terminated.
make[3]: *** [src/lib/tzcode/CMakeFiles/tzcode.dir/build.make:90: src/lib/tzcode/CMakeFiles/tzcode.dir/strptime.c.o] Error 1
make[3]: *** Waiting for unfinished jobs....
[  0%] Generating module.h
[  0%] No update step for 'bundled-nghttp2-project'
[  0%] Building C object CMakeFiles/zstd.dir/third_party/zstd/lib/common/debug.c.o
make[3]: Entering directory '/build/src/tarantool-2.10.2'
/build/src/tarantool-2.10.2/src/lib/tzcode/timelocal.c:36:10: fatal error: sys/cdefs.h: No such file or directory
   36 | #include <sys/cdefs.h>
      |          ^~~~~~~~~~~~~
compilation terminated.
make[3]: *** [src/lib/tzcode/CMakeFiles/tzcode.dir/build.make:104: src/lib/tzcode/CMakeFiles/tzcode.dir/timelocal.c.o] Error 1
```

Looks like the alpine Docker image needs to have some packages installed. Or I am missing something?

**Comment by Totktonada** [comment]
I would highligh a few things we discussed with Sergey privately:

* It would be good to move a version adjustment to Apline-specific part (`pack/apk.mk`) and name the adjusted version somehow like `APK_VERSION`. We're doing something similar with `DEB_VERSION` in `pack/deb.mk`.
* It would be good to have some more information about the keys (maybe just a link if it'll answer all questions). I don't understand what will going wrong without the patch and how a build time generated keys may be used for validation of packages in a repository.

**Comment by Totktonada** [comment]
> * It would be good to move a version adjustment to Apline-specific part (`pack/apk.mk`) and name the adjusted version somehow like `APK_VERSION`. We're doing something similar with `DEB_VERSION` in `pack/deb.mk`.

Discussed with Sergey and found that tarantool's build script sets a `VERSION` variable that is incorrect from `abuild`/`apk` point of view. Should be solved in https://github.com/tarantool/tarantool/pull/7942.

> * It would be good to have some more information about the keys (maybe just a link if it'll answer all questions). I don't understand what will going wrong without the patch and how a build time generated keys may be used for validation of packages in a repository.

Still unclear for me. No explanation was added into the pull request. If the patch will go through me, I must understand the usage scenario.

File: pack/apk.mk [hunk 1]
```diff
@@ -39,11 +40,13 @@ package: $(BUILDDIR)/$(TARBALL) \
 	@echo "Building APK packages"
 	@echo "-------------------------------------------------------------------"
 	sudo apk update
+	sudo chown -R ${USER}:${USER} /home/${USER}/.abuild/
```
Comment (Totktonada) [inline pack/apk.mk:43]: The assumption that the primary group is named the same as the user may be wrong. For example, my system:

```shell
$ id -nu
alex
$ id -ng
users
```

(IDs are different too.)

File: pack/apk.mk [hunk 2]
```diff
@@ -39,11 +40,13 @@ package: $(BUILDDIR)/$(TARBALL) \
...
 	rm -f $(BUILDDIR)/*/APKINDEX.tar.gz
-	rm -rf $(BUILDDIR)/src
+	rm -rf $(BUILDDIR)/src
```
Comment (Totktonada) [inline pack/apk.mk:45]: Trailing whitespaces, unnecessary change.

Comment (foxzi) [reply pack/apk.mk:45]: OK

File: pack/apk.mk (hunk 1 above)
Comment (foxzi) [reply pack/apk.mk:43]: OK, I replace it with changing owner only.

**Comment by Totktonada** [comment]
@foxzi There is no point in re-requesting review without taking into account my existing comments.

**Comment by Totktonada** [comment]
We agreed with Sergey that there is no point in signing a package with just generated key. All packages in a repository should be signed with the same key. The public part of the key is assumed to be available for users. (I guess a long lived repository may have several old keys and one actual, which is used for new packages. It doesn't change the package building pipeline -- there is one key to sign new packages.)

packpack should decline an attempt to create an Apline package without providing a key to sign the package. A proper error message and a link to further steps (to a readme section, for example) should be shown for a user in the case.

Sergey will elaborate the most convenient way(s) to pass the key into packpack for manual usage and calling from CI. I see the following variants:

* Pass the value of the key using an environment variable.
* Pass a file/directory path to copy into the container using an environment variable (copy, not binding a volume to don't touch the files on the host system!). Maybe with some default file/directory.

(I don't restrict it by those ways, it is just how I see it. And maybe one is better for hands, but another is better for automation -- so it'll worth to make both.)

AFAIU we need only private key for signing the package (and we don't need the public key). Correct me if I'm wrong. But if it is so, we can pass a file path (not a directory path) if we'll go by the second variant.

I'm waiting for results of Sergey's investivation. Please, re-request the review afterwards.

File: README.md [hunk 3]
```diff
@@ -317,6 +317,8 @@ see an example in [Tarantool GitHub](https://github.com/tarantool/tarantool) rep
 * `CCACHE*` - Config variables for ccache, such as CCACHE_DISABLE
 * `PRESERVE_ENVVARS` - a comma separated list of environment variables to
   preserve.
+* `ABUILD_KEY` - a private key for sign alpine packages generated by 
+  abuild-keygen
```
Comment (NickVolynkin) [inline README.md:321]:
```suggestion
* `ABUILD_KEY` - a private key for signing Alpine packages generated by 
  `abuild-keygen`.
```

File: packpack [hunk 4]
```diff
@@ -151,6 +151,13 @@ env | grep -E "^PRODUCT=|^VERSION=|^RELEASE=|^ABBREV=" > ${BUILDDIR}/env
 env | grep -E "^TARBALL_|^CHANGELOG_|^CCACHE_|^PACKAGECLOUD_" >> ${BUILDDIR}/env
 env | grep -E "^SMPFLAGS=|^OS=|^DIST=" >> ${BUILDDIR}/env
 
+#
+# Pass alpine build ssl key to docker for sign packages
```
Comment (NickVolynkin) [inline packpack:155]:
```suggestion
# Pass the Alpine build SSL key to Docker for signing packages
```

File: README.md (hunk 3 above)
Comment (foxzi) [reply README.md]: fixed

File: packpack (hunk 4 above)
Comment (foxzi) [reply packpack]: fixed

File: README.md [hunk 5]
```diff
@@ -317,6 +317,8 @@ see an example in [Tarantool GitHub](https://github.com/tarantool/tarantool) rep
 * `CCACHE*` - Config variables for ccache, such as CCACHE_DISABLE
 * `PRESERVE_ENVVARS` - a comma separated list of environment variables to
   preserve.
+* `ABUILD_KEY` - a private key for signing alpine packages generated by 
+  `abuild-keygen`
```
Comment (Totktonada) [inline README.md:321]: AFAIS, it is base64 encoded, but the readme doesn't say it.

Comment (Totktonada) [reply README.md:321]: BTW, your one-line instruction how to create the key (from the commit message) would be useful here.

File: packpack [hunk 6]
```diff
@@ -136,6 +135,7 @@ else
         echo "usermod -a -G sudo $USER 2>/dev/null || :";
         echo "usermod -a -G wheel $USER 2>/dev/null || :";
         echo "usermod -a -G adm $USER 2>/dev/null || :";
+        printf 'printf "%s\\tALL=(ALL)\\tNOPASSWD: ALL\\n" >> /etc/sudoers\n' $USER
```
Comment (Totktonada) [inline packpack:138]: Comes from a119cafc9d64ad346b9ac1face1625564414c34d? It would be good to rebase on top of the recent master.

File: README.md (hunk 5 above)
Comment (Totktonada) [reply README.md:321]: > AFAIS, it is base64 encoded, but the readme doesn't say it.

I see now, it is not so. It is encoded before writting to a file. Got it.

Comment (Totktonada) [reply README.md:321]: I'll add the comment about the instruction as a separate comment and will close this one.

File: pack/apk.mk [hunk 7]
```diff
@@ -39,11 +44,15 @@ package: $(BUILDDIR)/$(TARBALL) \
...
 	sudo apk update
+	mkdir ~/.abuild/
+	echo "PACKAGER_PRIVKEY=/home/${USER}/.abuild/private.rsa" > ~/.abuild/abuild.conf
+	echo "$(ABUILD_KEY)" | base64 -d > ~/.abuild/private.rsa
 	rm -f $(BUILDDIR)/*/APKINDEX.tar.gz
 	rm -rf $(BUILDDIR)/src
 	cd $(BUILDDIR); abuild checksum
 	cd $(BUILDDIR); abuild -r -P $(BUILDDIR) -s $(BUILDDIR)
 	mv -f $(BUILDDIR)/*/*.apk $(BUILDDIR)
+	tar -czf $(BUILDDIR)/keys.tgz -C /home/${USER}/.abuild/ .
```
Comment (Totktonada) [inline pack/apk.mk:55]: Is it useful, when a caller is already obligated to have the key? Or it was made for a previous version of the patch, when the keys were autogenerated and useless now? If the latter is true, I would remove it. Less possibilities to make a mistake and reveal the private key.

File: README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (foxzi) [reply README.md]: done

File: packpack (hunk 6 above)
Comment (foxzi) [reply packpack]: done

File: pack/apk.mk (hunk 1 above)
Comment (Totktonada) [reply pack/apk.mk]: Irrelevant now.
